### PR TITLE
Abyss Planck keymap

### PR DESCRIPTION
**Pull request by hein946**
# Abyss Planck keymap

## Function
- Assign ctrl+Space to the LOWER and RAISE keys for full-width conversion.
   - CONV_RAISE
   - CONV_LOWER

- Added the ability to switch layers for Mac and Windows, and changed it so that you can use the same actions such as copying.
(I'd really like to auto-detect and switch between Win and Mac...)
   - CHG_LAYER

### CODE REVIEW

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    case WIN:
+      if (record->event.pressed) {
+        print("mode just switched to WIN and this is a huge string\n");
+        set_single_persistent_default_layer(_WIN);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:217]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    case MAC:
+      if (record->event.pressed) {
+        set_single_persistent_default_layer(_MAC);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:223]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    case COLEMAK:
+      if (record->event.pressed) {
+        set_single_persistent_default_layer(_COLEMAK);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:229]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    case DVORAK:
+      if (record->event.pressed) {
+        set_single_persistent_default_layer(_DVORAK);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:235]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+          writePinLow(E6);
+        #endif
+      } else {
+        unregister_code(KC_LCTL);
+        #ifdef KEYBOARD_planck_rev5
+          writePinHigh(E6);
+        #endif
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:252]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+        layer_on(_PLOVER);
+        if (!eeconfig_is_enabled()) {
+            eeconfig_init();
+        }
+        keymap_config.raw = eeconfig_read_keymap();
+        keymap_config.nkro = 1;
+        eeconfig_update_keymap(keymap_config.raw);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:271]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+    case EXT_PLV:
+      if (record->event.pressed) {
+        #ifdef AUDIO_ENABLE
+          PLAY_SONG(plover_gb_song);
+        #endif
+        layer_off(_PLOVER);
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:280]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        layer_on(_LOWER);
+      }else{
+        layer_off(_LOWER);
+        if(timer_elapsed(key_timer) <= 130){
+          tap_code16(LCTL(KC_SPC));
+        }
+      }
+      return false;
+      break;
+
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:293]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        layer_on(_RAISE);
+      }else{
+        layer_off(_RAISE);
+        if(timer_elapsed(key_timer) <= 130){
+          tap_code16(LCTL(KC_SPC));
+        }
+      }
+      return false;
+      break;
+
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:306]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+          layer_invert(_MAC);
+          set_single_persistent_default_layer(_MAC);
+        }else{
+          layer_off(_MAC);
+          layer_invert(_WIN);
+          set_single_persistent_default_layer(_WIN);
+        }
+      }
+      return false;
+      break;
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/keymap.c:320]:
```suggestion
```

File: keyboards/planck/keymaps/abyss/config.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+   - etc.
+*/
+//#define MIDI_ADVANCED
+
+/* override number of MIDI tone keycodes (each octave adds 12 keycodes and allocates 12 bytes) */
+//#define MIDI_TONE_KEYCODE_OCTAVES 2
+
+// Most tactile encoders have detents every 4 stages
+#define ENCODER_RESOLUTION 4
+
```
Comment (fauxpark) [inline keyboards/planck/keymaps/abyss/config.h:39]:
```suggestion
```

**Comment by drashna** [comment]
@fauxpark I disagree with removing `break`.  I would rather have the `return false` removed.  But that's just my preference. 

@hein946 It looks like the only difference between the MAC and WIN layers is the position of Control and GUI.  If that's the case, you can remove one of the layers, and use the "Control GUI swap" keycodes of magic/bootmagic. 
https://docs.qmk.fm/#/feature_bootmagic?id=keycodes

Specifically, `CG_SWAP`, `CG_NORM`, but mostly `CG_TOGG`. 

These are persistent and available even if bootmagic is disabled (but not if magic is disabled). 


Additionally, we generally prefer that the name for the keymap match your github username.  It's not a hard requirement, but we highly recommend it.

**Comment by fauxpark** [comment]
Removing the `break`s keeps the current functionality. If it still works the same or better the other way around instead, go for it.

File: keyboards/planck/keymaps/abyss/keymap.c [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+ * | Brite| Ctrl | Alt  | GUI  |Lower | Space| Enter|Raise | Left | Down |  Up  |Right |
+ * `-----------------------------------------------------------------------------------'
+ */
+[_MAC] = LAYOUT_planck_grid(
+    KC_ESC,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_BSPC,
+    KC_TAB,  KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,
+    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSHIFT ,
+    BACKLIT, KC_LCTL, KC_LALT, KC_LGUI, CONV_LOWER,   KC_SPC,  KC_ENT, CONV_RAISE,   KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT
+),
+
```
Comment (drashna) [inline keyboards/planck/keymaps/abyss/keymap.c:87]: Just a heads up. It looks like the only real change here is that the Control and GUI keys are swapped. 

You can use the Bootmagic keycodes to swap these. https://docs.qmk.fm/#/feature_bootmagic?id=keycodes

The bootmagic keycodes work even if bootmagic is disabled.  And the CG_SWAP, CG_NORM, and CG_TOGG keycodes

**Pull request closed without merging**